File: next_project/cpp/src/topology.cpp
#include "topology.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <stdexcept>

namespace sim {

namespace {

constexpr double kPi = 3.14159265358979323846;
constexpr double kSigmaSquared = 4.0;

double offdiag_norm_sq(const std::vector<std::vector<double>>& m) {
    double acc = 0.0;
    for (std::size_t i = 0; i < m.size(); ++i) {
        for (std::size_t j = 0; j < m[i].size(); ++j) {
            if (i == j) continue;
            acc += m[i][j] * m[i][j];
        }
    }
    return acc;
}

}  // namespace

FormationTopology::FormationTopology(int num_followers, double spacing, double arm_length)
    : num_followers_(num_followers), spacing_(spacing), arm_length_(arm_length) {
    if (num_followers_ < 1) {
        throw std::invalid_argument("num_followers must be >= 1");
    }
    if (spacing_ <= 0.0) {
        throw std::invalid_argument("spacing must be positive");
    }

    current_offsets_ = compute_offsets("v_shape");
}

std::vector<Vec3> FormationTopology::compute_v_shape() const {
    std::vector<Vec3> offsets;
    offsets.reserve(static_cast<std::size_t>(num_followers_));

    for (int i = 0; i < num_followers_; ++i) {
        const int rank = (i / 2) + 1;
        const int side = (i % 2 == 0) ? -1 : 1;
        offsets.emplace_back(-rank * spacing_, side * rank * spacing_, 0.0);
    }
    return offsets;
}

std::vector<Vec3> FormationTopology::compute_diamond() const {
    std::vector<Vec3> offsets{
        Vec3{-spacing_, -spacing_, 0.0},
        Vec3{-spacing_, spacing_, 0.0},
        Vec3{-2.0 * spacing_, 0.0, 0.0},
    };

    while (static_cast<int>(offsets.size()) < num_followers_) {
        const int extra_rank = static_cast<int>(offsets.size()) - 2;
        offsets.emplace_back(-2.0 * spacing_ - extra_rank * spacing_, 0.0, 0.0);
    }

    offsets.resize(static_cast<std::size_t>(num_followers_));
    return offsets;
}

std::vector<Vec3> FormationTopology::compute_line() const {
    std::vector<Vec3> offsets;
    offsets.reserve(static_cast<std::size_t>(num_followers_));

    for (int i = 0; i < num_followers_; ++i) {
        offsets.emplace_back(-(i + 1) * spacing_, 0.0, 0.0);
    }

    return offsets;
}

std::vector<Vec3> FormationTopology::compute_triangle() const {
    std::vector<Vec3> offsets;
    offsets.reserve(static_cast<std::size_t>(num_followers_));

    const double row_height = spacing_ * std::sqrt(3.0) / 2.0;
    int placed = 0;
    int row = 1;

    while (placed < num_followers_) {
        const int num_in_row = row + 1;
        const double row_y_start = -row * spacing_ / 2.0;

        for (int col = 0; col < num_in_row && placed < num_followers_; ++col) {
            const double dx = -row * row_height;
            const double dy = row_y_start + col * spacing_;
            offsets.emplace_back(dx, dy, 0.0);
            ++placed;
        }

        ++row;
    }

    return offsets;
}

std::vector<Vec3> FormationTopology::compute_offsets(const std::string& formation) const {
    if (formation == "v_shape") {
        return compute_v_shape();
    }
    if (formation == "diamond") {
        return compute_diamond();
    }
    if (formation == "line") {
        return compute_line();
    }
    if (formation == "triangle") {
        return compute_triangle();
    }
    if (formation == "custom") {
        if (custom_offsets_.empty()) {
            throw std::invalid_argument("custom formation requires set_custom_offsets first");
        }
        if (static_cast<int>(custom_offsets_.size()) != num_followers_) {
            throw std::invalid_argument("custom offsets count does not match followers");
        }
        return custom_offsets_;
    }

    throw std::invalid_argument("unknown formation: " + formation);
}

std::vector<Vec3> FormationTopology::get_offsets(const std::string& formation) const {
    return compute_offsets(formation);
}

void FormationTopology::set_custom_offsets(const std::vector<Vec3>& offsets) {
    if (static_cast<int>(offsets.size()) != num_followers_) {
        throw std::invalid_argument("offset count does not match followers");
    }
    custom_offsets_ = offsets;
}

void FormationTopology::switch_formation(const std::string& target_formation, double transition_time, double current_time) {
    if (transition_time <= 0.0) {
        throw std::invalid_argument("transition_time must be positive");
    }

    source_offsets_ = current_offsets_;
    target_offsets_ = compute_offsets(target_formation);
    target_formation_ = target_formation;
    transition_time_ = transition_time;
    switch_start_time_ = 0.0;  // 对齐 Python：瞬时切换
    switching_ = true;
}

std::vector<Vec3> FormationTopology::get_current_offsets(double elapsed_time) {
    if (!switching_) {
        return current_offsets_;
    }

    const double t = elapsed_time - switch_start_time_;
    if (t >= transition_time_) {
        current_offsets_ = target_offsets_;
        current_formation_ = target_formation_;
        switching_ = false;
        return current_offsets_;
    }

    const double alpha = 0.5 * (1.0 - std::cos(kPi * t / transition_time_));

    std::vector<Vec3> blended;
    blended.reserve(current_offsets_.size());

    for (std::size_t i = 0; i < source_offsets_.size(); ++i) {
        blended.push_back((1.0 - alpha) * source_offsets_[i] + alpha * target_offsets_[i]);
    }

    return blended;
}

double FormationTopology::envelope_radius(const std::string& formation) const {
    auto [lateral, longitudinal, vertical] = envelope_per_axis(formation);
    return std::max({lateral, longitudinal, vertical});
}

std::tuple<double, double, double> FormationTopology::envelope_per_axis(const std::string& formation) const {
    const auto name = formation.empty()
        ? (switching_ ? target_formation_ : current_formation_)
        : formation;
    const auto offsets = get_offsets(name);
    if (offsets.empty()) {
        return {arm_length_, arm_length_, arm_length_};
    }
    double max_abs_x = 0.0;
    double max_abs_y = 0.0;
    double max_abs_z = 0.0;
    for (const auto& off : offsets) {
        max_abs_x = std::max(max_abs_x, std::abs(off.x));
        max_abs_y = std::max(max_abs_y, std::abs(off.y));
        max_abs_z = std::max(max_abs_z, std::abs(off.z));
    }
    return {
        max_abs_y + arm_length_,
        max_abs_x + arm_length_,
        max_abs_z + arm_length_,
    };
}

bool FormationTopology::auto_shrink(const std::tuple<double, double, double>& channel_width,
                                    const std::tuple<double, double, double>* envelope) {
    const auto env = envelope ? *envelope : envelope_per_axis(current_formation_);
    const bool too_narrow =
        std::get<0>(channel_width) < 2.0 * std::get<0>(env)
        || std::get<1>(channel_width) < 2.0 * std::get<1>(env)
        || std::get<2>(channel_width) < 2.0 * std::get<2>(env);
    const bool roomy =
        std::get<0>(channel_width) > 3.0 * std::get<0>(env)
        && std::get<1>(channel_width) > 3.0 * std::get<1>(env)
        && std::get<2>(channel_width) > 3.0 * std::get<2>(env);
    if (too_narrow && current_formation_ != "line") {
        switch_formation("line", 1.5, 0.0);
        current_offsets_ = target_offsets_;
        current_formation_ = target_formation_;
        switching_ = false;
        return true;
    }
    if (roomy && current_formation_ == "line") {
        switch_formation("diamond", 2.0, 0.0);
        current_offsets_ = target_offsets_;
        current_formation_ = target_formation_;
        switching_ = false;
        return true;
    }
    return false;
}

std::string FormationTopology::fault_reconfigure(const std::vector<int>& failed_indices,
                                                 double transition_time,
                                                 double current_time) {
    std::string best = TopologyGraph::best_reconfig_topology(
        failed_indices, num_followers_, spacing_);
    if (!best.empty()) {
        switch_formation(best, transition_time, current_time);
    }
    return best;
}

TopologyGraph::TopologyGraph(const std::vector<Vec3>& offsets)
    : offsets_(offsets) {}

double TopologyGraph::algebraic_connectivity() const {
    if (offsets_.empty()) return 0.0;
    std::vector<Vec3> positions{Vec3{}};
    positions.insert(positions.end(), offsets_.begin(), offsets_.end());
    const int n = static_cast<int>(positions.size());
    std::vector<std::vector<double>> laplacian(static_cast<std::size_t>(n), std::vector<double>(static_cast<std::size_t>(n), 0.0));

    for (int i = 0; i < n; ++i) {
        for (int j = i + 1; j < n; ++j) {
            const Vec3 delta = positions[i] - positions[j];
            const double d2 = dot(delta, delta);
            const double w = std::exp(-d2 / kSigmaSquared);
            laplacian[static_cast<std::size_t>(i)][static_cast<std::size_t>(i)] += w;
            laplacian[static_cast<std::size_t>(j)][static_cast<std::size_t>(j)] += w;
            laplacian[static_cast<std::size_t>(i)][static_cast<std::size_t>(j)] -= w;
            laplacian[static_cast<std::size_t>(j)][static_cast<std::size_t>(i)] -= w;
        }
    }

    const int max_iter = 64;
    for (int iter = 0; iter < max_iter && offdiag_norm_sq(laplacian) > 1e-18; ++iter) {
        int p = 0;
        int q = 1;
        double max_val = 0.0;
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                const double v = std::abs(laplacian[static_cast<std::size_t>(i)][static_cast<std::size_t>(j)]);
                if (v > max_val) {
                    max_val = v;
                    p = i;
                    q = j;
                }
            }
        }
        if (max_val < 1e-12) {
            break;
        }

        const double app = laplacian[static_cast<std::size_t>(p)][static_cast<std::size_t>(p)];
        const double aqq = laplacian[static_cast<std::size_t>(q)][static_cast<std::size_t>(q)];
        const double apq = laplacian[static_cast<std::size_t>(p)][static_cast<std::size_t>(q)];
        const double tau = (aqq - app) / (2.0 * apq);
        const double t = ((tau >= 0.0) ? 1.0 : -1.0) / (std::abs(tau) + std::sqrt(1.0 + tau * tau));
        const double c = 1.0 / std::sqrt(1.0 + t * t);
        const double s = t * c;

        for (int k = 0; k < n; ++k) {
            if (k == p || k == q) continue;
            const double aik = laplacian[static_cast<std::size_t>(k)][static_cast<std::size_t>(p)];
            const double akq = laplacian[static_cast<std::size_t>(k)][static_cast<std::size_t>(q)];
            laplacian[static_cast<std::size_t>(k)][static_cast<std::size_t>(p)] = c * aik - s * akq;
            laplacian[static_cast<std::size_t>(p)][static_cast<std::size_t>(k)] = laplacian[static_cast<std::size_t>(k)][static_cast<std::size_t>(p)];
            laplacian[static_cast<std::size_t>(k)][static_cast<std::size_t>(q)] = s * aik + c * akq;
            laplacian[static_cast<std::size_t>(q)][static_cast<std::size_t>(k)] = laplacian[static_cast<std::size_t>(k)][static_cast<std::size_t>(q)];
        }

        const double new_app = c * c * app - 2.0 * s * c * apq + s * s * aqq;
        const double new_aqq = s * s * app + 2.0 * s * c * apq + c * c * aqq;
        laplacian[static_cast<std::size_t>(p)][static_cast<std::size_t>(p)] = new_app;
        laplacian[static_cast<std::size_t>(q)][static_cast<std::size_t>(q)] = new_aqq;
        laplacian[static_cast<std::size_t>(p)][static_cast<std::size_t>(q)] = 0.0;
        laplacian[static_cast<std::size_t>(q)][static_cast<std::size_t>(p)] = 0.0;
    }

    std::vector<double> eigenvals;
    eigenvals.reserve(static_cast<std::size_t>(n));
    for (int i = 0; i < n; ++i) {
        eigenvals.push_back(laplacian[static_cast<std::size_t>(i)][static_cast<std::size_t>(i)]);
    }
    std::sort(eigenvals.begin(), eigenvals.end());
    return eigenvals.size() >= 2 ? eigenvals[1] : 0.0;
}

std::string TopologyGraph::best_reconfig_topology(const std::vector<int>& failed_indices,
                                                  int num_followers,
                                                  double spacing,
                                                  const std::vector<std::string>& candidates) {
    int healthy = num_followers - static_cast<int>(failed_indices.size());
    if (healthy < 1) return "";

    std::string best;
    double best_score = -1.0;
    for (const auto& formation : candidates) {
        auto offsets = compute_offsets_for(formation, healthy, spacing);
        if (static_cast<int>(offsets.size()) != healthy) continue;
        TopologyGraph graph(offsets);
        double score = graph.algebraic_connectivity();
        if (score > best_score) {
            best_score = score;
            best = formation;
        }
    }
    return best;
}

std::vector<Vec3> TopologyGraph::compute_offsets_for(const std::string& formation,
                                                     int num_followers,
                                                     double spacing) {
    FormationTopology topo(num_followers, spacing);
    return topo.get_offsets(formation);
}

}  // namespace sim

File: next_project/core/topology.py
"""编队拓扑管理。

用途
----
管理无人机编队的相对偏移量，并支持队形之间的平滑切换。

原理
----
设当前偏移为 Delta_src，目标偏移为 Delta_tgt，过渡时间为 T。
在 t in [0, T] 内使用余弦插值：

    alpha(t) = 0.5 * (1 - cos(pi * t / T))
    Delta(t) = (1 - alpha) * Delta_src + alpha * Delta_tgt

这种方式在起点和终点均具有零速度特性，可减小切换冲击。


"""

from __future__ import annotations

import numpy as np


class FormationTopology:
    """动态编队拓扑管理器。"""

    PREDEFINED_FORMATIONS = ("v_shape", "diamond", "line", "triangle")

    def __init__(self, num_followers: int = 3, spacing: float = 2.0, arm_length: float = 0.2):
        if num_followers < 1:
            raise ValueError("num_followers must be >= 1")
        if spacing <= 0:
            raise ValueError("spacing must be positive")
        self.num_followers = int(num_followers)
        self.spacing = float(spacing)
        self.arm_length = float(arm_length)
        self.current_formation = "v_shape"
        self._current_offsets = self._compute_offsets("v_shape")
        self.custom_offsets = None
        self._switching = False
        self._switch_start_time = 0.0
        self._transition_time = 0.0
        self._source_offsets = []
        self._target_offsets = []
        self._target_formation = ""

    def _compute_v_shape(self):
        offsets = []
        for i in range(self.num_followers):
            rank = (i // 2) + 1
            side = -1 if i % 2 == 0 else 1
            offsets.append(np.array([-rank * self.spacing, side * rank * self.spacing, 0.0], dtype=float))
        return offsets

    def _compute_diamond(self):
        offsets = [
            np.array([-self.spacing, -self.spacing, 0.0], dtype=float),
            np.array([-self.spacing, self.spacing, 0.0], dtype=float),
            np.array([-2.0 * self.spacing, 0.0, 0.0], dtype=float),
        ]
        while len(offsets) < self.num_followers:
            extra_rank = len(offsets) - 2
            offsets.append(np.array([-2.0 * self.spacing - extra_rank * self.spacing, 0.0, 0.0], dtype=float))
        return offsets[:self.num_followers]

    def _compute_line(self):
        return [np.array([-(i + 1) * self.spacing, 0.0, 0.0], dtype=float) for i in range(self.num_followers)]

    def _compute_triangle(self):
        offsets = []
        row_height = self.spacing * np.sqrt(3.0) / 2.0
        placed = 0
        row = 1
        while placed < self.num_followers:
            num_in_row = row + 1
            row_y_start = -row * self.spacing / 2.0
            for col in range(num_in_row):
                if placed >= self.num_followers:
                    break
                dx = -row * row_height
                dy = row_y_start + col * self.spacing
                offsets.append(np.array([dx, dy, 0.0], dtype=float))
                placed += 1
            row += 1
        return offsets

    def _compute_offsets(self, formation: str):
        if formation == "v_shape":
            return self._compute_v_shape()
        if formation == "diamond":
            return self._compute_diamond()
        if formation == "line":
            return self._compute_line()
        if formation == "triangle":
            return self._compute_triangle()
        if formation == "custom":
            if self.custom_offsets is None:
                raise ValueError("custom formation requires set_custom_offsets first")
            if len(self.custom_offsets) != self.num_followers:
                raise ValueError("custom offsets count does not match followers")
            return [off.copy() for off in self.custom_offsets]
        raise ValueError(f"unknown formation: {formation}")

    def get_offsets(self, formation: str):
        return self._compute_offsets(formation)

    def set_custom_offsets(self, offsets) -> None:
        if len(offsets) != self.num_followers:
            raise ValueError("offset count does not match followers")
        self.custom_offsets = [np.array(off, dtype=float).flatten()[:3] for off in offsets]

    def switch_formation(self, target_formation: str, transition_time: float = 5.0) -> None:
        if transition_time <= 0:
            raise ValueError("transition_time must be positive")
        target_offsets = self._compute_offsets(target_formation)
        self._source_offsets = [off.copy() for off in self._current_offsets]
        self._target_offsets = target_offsets
        self._target_formation = target_formation
        self._transition_time = float(transition_time)
        self._switch_start_time = 0.0
        self._switching = True

    def get_current_offsets(self, elapsed_time: float):
        if not self._switching:
            return [off.copy() for off in self._current_offsets]
        t = elapsed_time - self._switch_start_time
        if t >= self._transition_time:
            self._current_offsets = [off.copy() for off in self._target_offsets]
            self.current_formation = self._target_formation
            self._switching = False
            return [off.copy() for off in self._current_offsets]
        alpha = 0.5 * (1.0 - np.cos(np.pi * t / self._transition_time))
        return [(1.0 - alpha) * src + alpha * tgt for src, tgt in zip(self._source_offsets, self._target_offsets)]

    def envelope_radius(self, formation_type: str | None = None) -> float:
        """返回当前/指定队形下编队完整包络半径（含臂长）。

        用途
        ----
        供路径规划器计算膨胀半径：r_inflate = envelope_radius + safety_margin。
        envelope_radius 已包含臂长，调用方不应再次叠加 arm_length。

        原理
        ----
        遍历队形偏移量，取最大范数并加上臂长。
        """
        offsets = self.get_offsets(formation_type or self.current_formation)
        max_dist = max(float(np.linalg.norm(off)) for off in offsets) if offsets else 0.0
        return max_dist + self.arm_length

    @property
    def is_switching(self) -> bool:
        return self._switching

    # ------------------------------------------------------------------
    # 论文4: 容错拓扑重构
    # ------------------------------------------------------------------

    def fault_reconfigure(self, failed_indices: list[int], transition_time: float = 3.0) -> str | None:
        """故障后自动选择最佳拓扑并开始切换。

        failed_indices: 故障从机索引列表（0-based）。
        返回选中的拓扑类型，或 None（无可用拓扑）。
        """
        candidates = ["v_shape", "diamond", "line", "triangle"]
        best_topo = TopologyGraph.best_reconfig_topology(
            failed_indices, self.num_followers, self.spacing, candidates
        )
        if best_topo is not None:
            self.switch_formation(best_topo, transition_time)
        return best_topo

    # ------------------------------------------------------------------
    # 论文2: 队形自动收缩
    # ------------------------------------------------------------------

    def auto_shrink(self, channel_width: float, envelope: float | None = None) -> bool:
        """基于通道宽度自动触发队形缩放。

        若 channel_width < 2*envelope → 缩为 line 队形。
        返回 True 表示已触发切换。
        """
        if envelope is None:
            envelope = self.envelope_radius(self.current_formation)
        if channel_width < 2.0 * envelope and self.current_formation != "line":
            self.switch_formation("line", transition_time=1.5)
            return True
        elif channel_width > 3.0 * envelope and self.current_formation == "line":
            self.switch_formation("diamond", transition_time=2.0)
            return True
        return False


class TopologyGraph:
    """编队图论拓扑模型（论文4 L3）。

    用途
    ----
    用图 Laplacian 矩阵描述编队拓扑关系，提取代数连通度 λ₂，
    作为拓扑鲁棒性度量和故障重构选择的依据。
    """

    def __init__(self, offsets: list[np.ndarray]):
        self.n = len(offsets) + 1  # followers + leader
        self._offsets = [np.asarray(o, dtype=float) for o in offsets]
        self._laplacian = self._build_laplacian()

    def _build_laplacian(self) -> np.ndarray:
        """从几何偏移构建 Laplacian 矩阵。

        节点 0 = 领航机，节点 1..n = 从机。
        a_ij = exp(-||offset_i - offset_j||² / σ²)
        """
        n = self.n
        sigma = 2.0  # 距离缩放参数
        positions = [np.zeros(3, dtype=float)] + self._offsets
        A = np.zeros((n, n), dtype=float)
        for i in range(n):
            for j in range(i + 1, n):
                d2 = float(np.sum((positions[i] - positions[j]) ** 2))
                w = np.exp(-d2 / (sigma ** 2))
                A[i, j] = w
                A[j, i] = w
        D = np.diag(np.sum(A, axis=1))
        return D - A

    @property
    def algebraic_connectivity(self) -> float:
        """λ₂ = Laplacian 的第二小特征值。越大表示拓扑连通性越强。"""
        eigenvals = np.linalg.eigvalsh(self._laplacian)
        return float(eigenvals[1])

    @property
    def laplacian(self) -> np.ndarray:
        return self._laplacian

    @staticmethod
    def best_reconfig_topology(
        failed_indices: list[int],
        num_followers: int,
        spacing: float,
        candidates: list[str] = ("v_shape", "diamond", "line", "triangle"),
    ) -> str | None:
        """故障后选择 λ₂ 最大的可行拓扑。

        1. 排除含故障机的候选配置（剩余健康的从机数量要能组成该队形）
        2. 对每个候选拓扑构建 Laplacian
        3. 返回 λ₂ 最大的拓扑类型
        """
        healthy = num_followers - len(failed_indices)
        if healthy < 1:
            return None

        best = candidates[0]
        best_lambda2 = -1.0

        for topo in candidates:
            # 计算该队形对健康数量的偏移
            offsets = TopologyGraph._compute_offsets_for(topo, healthy, spacing)
            if offsets is None or len(offsets) != healthy:
                continue
            tg = TopologyGraph(offsets)
            l2 = tg.algebraic_connectivity
            if l2 > best_lambda2:
                best_lambda2 = l2
                best = topo

        return best

    @staticmethod
    def _compute_offsets_for(formation: str, n: int, spacing: float) -> list[np.ndarray] | None:
        """为指定队形和从机数计算偏移量（静态辅助方法）。"""
        try:
            ft = FormationTopology(num_followers=n, spacing=spacing)
            return ft._compute_offsets(formation)
        except Exception:
            return None
